src/ray/raylet/runtime_env_agent_client.cc

```diff
@@ -0,0 +1,395 @@
+                                          std::to_string(res_.result_int())));
+    }
+
+    // Gracefully close the socket
+    stream_.socket().shutdown(tcp::socket::shutdown_both, ec);
+    // not_connected happens sometimes so don't bother reporting it.
+    if (ec && ec != beast::errc::not_connected) {
+      RAY_LOG(INFO) << "on_read error after response body reseived: " << ec.message();
+    }
+  }
+};
+
+class HttpRuntimeEnvAgentClient : public RuntimeEnvAgentClient {
+ public:
+  HttpRuntimeEnvAgentClient(instrumented_io_context &io_context,
+                            const std::string &address,
+                            int port,
+                            std::function<std::shared_ptr<boost::asio::deadline_timer>(
+                                std::function<void()>, uint32_t delay_ms)> delay_executor)
+      : io_context_(io_context),
+        address_(address),
+        port_str_(std::to_string(port)),
+        delay_executor_(delay_executor) {}
+  ~HttpRuntimeEnvAgentClient() {}
+
+  template <typename T>
+  using SuccCallback = std::function<void(T)>;
+  using FailCallback = std::function<void(ray::Status)>;
+  template <typename T>
+  using TryInvokeOnce = std::function<void(SuccCallback<T>, FailCallback)>;
+
+  /// @brief Invoke `try_invoke_once`. If it fails with a retriable error, retries after
+  /// `agent_manager_retry_interval_ms` via the `delay_executor_`.
+  //
+  // "Retriable error" means `IsNotFound`.
+  /// @tparam T the return type on success.
+  /// @param try_invoke_once
+  /// @param succ_callback
+  /// @param fail_callback
+  template <typename T>
+  void RetryInvoke(TryInvokeOnce<T> try_invoke_once,
+                   SuccCallback<T> succ_callback,
+                   FailCallback fail_callback) {
+    try_invoke_once(succ_callback, [=](ray::Status status) {
+      if (status.IsNotFound()) {
+        RAY_LOG(INFO) << "Scheduling a retry in "
+                      << RayConfig::instance().agent_manager_retry_interval_ms()
+                      << "ms...";
+        this->delay_executor_(
+            [=]() { RetryInvoke(try_invoke_once, succ_callback, fail_callback); },
+            RayConfig::instance().agent_manager_retry_interval_ms());
+      } else {
+        // Non retryable errors, invoke fail_callback
+        fail_callback(std::move(status));
+      }
+    });
+  }
+
+  // Making HTTP call.
+  // POST /get_or_create_runtime_env
+  // Body = proto rpc::GetOrCreateRuntimeEnvRequest
+  void GetOrCreateRuntimeEnv(const JobID &job_id,
+                             const std::string &serialized_runtime_env,
+                             const rpc::RuntimeEnvConfig &runtime_env_config,
+                             const std::string &serialized_allocated_resource_instances,
+                             GetOrCreateRuntimeEnvCallback callback) override {
+    RetryInvoke<rpc::GetOrCreateRuntimeEnvReply>(
+        [=](SuccCallback<rpc::GetOrCreateRuntimeEnvReply> succ_callback,
```

Comment:
Consider moving some of potentially large nputs here instead? (e.g., serialized_runtime_env, runtime_env_config )

Reply:
the argument is const ref so we can't move, but don't copy either. In the session creation we did use `std::move(payload)` so only 2 copy happens (1 to add to request proto, 1 to serialize the proto)